File: sdk-js/src/pipeline.js
/**
 * Orchestration déclarative de plusieurs agents. Miroir de
 * `intermesh.pipeline` côté Python — voir ce module pour le contexte.
 *
 *   import { InterMeshPipeline } from 'intermesh';
 *
 *   const pipeline = new InterMeshPipeline(orchestrator)
 *     .step('Traduire', { capabilities: ['translate'] })
 *     .step('Calculer', {
 *       capabilities: ['calculate'],
 *       inputFn: (prev) => ({ expression: prev.translated_text }),
 *     });
 *   const result = await pipeline.run({ text: 'compute forty two' });
 */

export class PipelineError extends Error {}

class Step {
  constructor(title, { agent = null, capabilities = [], roles = [], metadata = {}, inputFn = null, timeout = 15.0 } = {}) {
    if (!agent && capabilities.length === 0 && roles.length === 0) {
      throw new PipelineError(
        `Étape '${title}' : précisez 'agent' ou un critère de découverte (capabilities/roles).`
      );
    }
    this.title = title;
    this.agent = agent;
    this.capabilities = capabilities;
    this.roles = roles;
    this.metadata = metadata;
    this.inputFn = inputFn;
    this.timeout = timeout;
  }
}

/**
 * Enchaîne des tâches sur plusieurs agents, chacun trouvé par capacité ou
 * nommé explicitement. La sortie d'une étape est l'entrée de la suivante,
 * via `inputFn` si la forme doit changer.
 */
export class InterMeshPipeline {
  constructor(orchestrator) {
    this.orchestrator = orchestrator;
    this._steps = [];
  }

  /**
   * Ajoute une étape et retourne le pipeline, pour enchaîner les appels.
   *
   * @param {string} title Titre de la tâche InterMesh, visible dans le dashboard.
   * @param {object} opts
   * @param {string} [opts.agent] Nom exact de l'agent visé. Omis, l'agent
   *   est recherché par `capabilities`/`roles` au moment de l'exécution —
   *   pas à la déclaration, donc un agent qui rejoint le réseau
   *   entre-temps compte.
   * @param {string[]} [opts.capabilities] Critère de découverte, toutes requises.
   * @param {string[]} [opts.roles] Critère de découverte, un seul suffit.
   * @param {(prev:*)=>*} [opts.inputFn] Reçoit la sortie de l'étape
   *   précédente (ou l'entrée initiale pour la première) et produit
   *   l'entrée de celle-ci. Par défaut, transmise telle quelle.
   * @param {number} [opts.timeout=15.0] Délai d'attente de cette étape, en secondes.
   */
  step(title, opts = {}) {
    this._steps.push(new Step(title, opts));
    return this;
  }

  async _resolveAgent(step) {
    if (step.agent) return step.agent;
    const found = await this.orchestrator.discover({
      capabilities: step.capabilities, roles: step.roles, metadata: step.metadata,
    });
    if (!found.count) {
      throw new PipelineError(
        `Étape '${step.title}' : aucun agent disponible pour ` +
        `capabilities=${JSON.stringify(step.capabilities)} roles=${JSON.stringify(step.roles)}.`
      );
    }
    return found.agents[0].name;
  }

  /**
   * Exécute les étapes dans l'ordre déclaré, en série.
   *
   * Une étape en échec interrompt le pipeline sans exécuter les
   * suivantes, plutôt que de continuer sur une sortie invalide ; l'erreur
   * de `submitTask` (timeout ou échec distant) se propage telle quelle.
   *
   * @returns {Promise<{output:*, history:Array<{title:string, agent:string, output:*}>}>}
   */
  async run(initialInput = null) {
    if (this._steps.length === 0) {
      throw new PipelineError("Le pipeline n'a aucune étape.");
    }

    let result = initialInput;
    const history = [];

    for (const step of this._steps) {
      const target = await this._resolveAgent(step);
      const payload = step.inputFn ? step.inputFn(result) : result;
      result = await this.orchestrator.submitTask(step.title, target, payload);
      history.push({ title: step.title, agent: target, output: result });
    }

    return { output: result, history };
  }
}

/**
 * Soumet plusieurs tâches en parallèle et attend toutes les réponses.
 *
 * Le pendant du pipeline séquentiel : là où `InterMeshPipeline` enchaîne,
 * `fanOut` interroge plusieurs agents à la fois.
 *
 *   const results = await fanOut(orchestrator, [
 *     ['fr', { region: 'FR' }],
 *     ['de', { region: 'DE' }],
 *   ], { agents: { fr: 'marché_fr', de: 'marché_de' } });
 *
 * @param {*} orchestrator Agent InterMesh dont `discover`/`submitTask` porte
 *   les branches.
 * @param {Array<[string, *]>} branches Paires `[clé, entrée]`. La clé
 *   identifie la branche dans le résultat.
 * @param {object} [opts]
 * @param {Record<string,string[]>} [opts.capabilities] Critère de
 *   découverte par clé de branche.
 * @param {Record<string,string>} [opts.agents] Nom d'agent exact par clé
 *   de branche — prime sur `capabilities` pour cette branche.
 * @returns {Promise<Record<string,*>>} Une branche en échec y figure
 *   comme l'erreur levée, jamais levée elle-même : les branches saines
 *   restent exploitables même si une autre a échoué.
 */
export async function fanOut(orchestrator, branches, { capabilities = {}, agents = {} } = {}) {
  const runBranch = async (key, payload) => {
    let target = agents[key];
    if (!target) {
      const found = await orchestrator.discover({ capabilities: capabilities[key] || [] });
      if (!found.count) throw new PipelineError(`Branche '${key}' : aucun agent disponible.`);
      target = found.agents[0].name;
    }
    return orchestrator.submitTask(key, target, payload);
  };

  const keys = branches.map(([k]) => k);
  const settled = await Promise.allSettled(branches.map(([k, payload]) => runBranch(k, payload)));

  const results = {};
  settled.forEach((outcome, i) => {
    results[keys[i]] = outcome.status === 'fulfilled' ? outcome.value : outcome.reason;
  });
  return results;
}

File: sdk-python/intermesh/pipeline.py
"""
Orchestration déclarative de plusieurs agents.

`discover()` et `submit_task()` suffisent à composer un workflow, comme le
montre `examples/agent_a.py` : mais chaque enchaînement y est réécrit à la
main — chercher l'agent, soumettre, extraire le résultat, recommencer.
`InterMeshPipeline` capture ce patron une fois pour toutes, et `fan_out`
capture le cas symétrique : plusieurs agents interrogés en parallèle.

    from intermesh import InterMeshPipeline

    pipeline = (
        InterMeshPipeline(orchestrator)
        .step("Traduire", capabilities=["translate"],
              input_fn=lambda _: {"text": "Compute forty two", "target_lang": "fr"})
        .step("Calculer", capabilities=["calculate"],
              input_fn=lambda prev: {"expression": prev["translated_text"]})
    )
    result = await pipeline.run()
    print(result.output)

Ni `InterMeshPipeline` ni `fan_out` ne parlent au Hub directement : ils
délèguent à l'agent orchestrateur qui les instancie, donc à la même
identité, au même jeton, au même chiffrement que le reste du SDK.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, Optional


class PipelineError(Exception):
    """Une étape n'a trouvé aucun agent, ou son exécution a échoué."""


@dataclass
class StepResult:
    """Trace d'une étape exécutée : utile pour déboguer un pipeline échoué."""
    title: str
    agent: str
    output: Any


@dataclass
class PipelineResult:
    """Sortie de la dernière étape, avec l'historique complet du parcours."""
    output: Any
    history: list[StepResult] = field(default_factory=list)


class _Step:
    def __init__(
        self,
        title: str,
        *,
        agent: Optional[str] = None,
        capabilities: Optional[Iterable[str]] = None,
        roles: Optional[Iterable[str]] = None,
        metadata: Optional[dict] = None,
        input_fn: Optional[Callable[[Any], Any]] = None,
        timeout: float = 15.0,
    ):
        if agent is None and not capabilities and not roles:
            raise PipelineError(
                f"Étape '{title}' : précisez `agent=` ou un critère de "
                f"découverte (`capabilities=`/`roles=`)."
            )
        self.title = title
        self.agent = agent
        self.capabilities = list(capabilities or [])
        self.roles = list(roles or [])
        self.metadata = metadata or {}
        self.input_fn = input_fn
        self.timeout = timeout


class InterMeshPipeline:
    """
    Enchaîne des tâches sur plusieurs agents, chacun trouvé par capacité
    ou nommé explicitement. La sortie d'une étape est l'entrée de la
    suivante, via `input_fn` si la forme doit changer.
    """

    def __init__(self, orchestrator):
        self.orchestrator = orchestrator
        self._steps: list[_Step] = []

    def step(
        self,
        title: str,
        *,
        agent: Optional[str] = None,
        capabilities: Optional[Iterable[str]] = None,
        roles: Optional[Iterable[str]] = None,
        metadata: Optional[dict] = None,
        input_fn: Optional[Callable[[Any], Any]] = None,
        timeout: float = 15.0,
    ) -> "InterMeshPipeline":
        """
        Ajoute une étape et retourne le pipeline, pour enchaîner les appels.

        Args:
            title:        Titre de la tâche InterMesh, visible dans le dashboard.
            agent:        Nom exact de l'agent visé. Omis, l'agent est
                          recherché par `capabilities`/`roles` au moment de
                          l'exécution — pas à la déclaration, donc un
                          agent qui rejoint le réseau entre-temps compte.
            capabilities: Critère de découverte, toutes requises.
            roles:        Critère de découverte, un seul suffit.
            input_fn:     Reçoit la sortie de l'étape précédente (ou
                          l'entrée initiale pour la première étape) et
                          produit l'entrée de celle-ci. Par défaut, la
                          sortie précédente est transmise telle quelle.
            timeout:      Délai d'attente de cette étape, en secondes.
        """
        self._steps.append(_Step(
            title, agent=agent, capabilities=capabilities, roles=roles,
            metadata=metadata, input_fn=input_fn, timeout=timeout,
        ))
        return self

    async def _resolve_agent(self, step: _Step) -> str:
        if step.agent:
            return step.agent
        found = await self.orchestrator.discover(
            capabilities=step.capabilities, roles=step.roles, metadata=step.metadata,
        )
        if found.get("count", 0) == 0:
            raise PipelineError(
                f"Étape '{step.title}' : aucun agent disponible pour "
                f"capabilities={step.capabilities} roles={step.roles}."
            )
        return found["agents"][0]["name"]

    async def run(self, initial_input: Any = None) -> PipelineResult:
        """
        Exécute les étapes dans l'ordre déclaré, en série.

        Raises:
            PipelineError: une étape n'a trouvé aucun agent.
            TimeoutError, RuntimeError: propagées telles quelles depuis
                `submit_task` — une étape en échec interrompt le pipeline
                sans exécuter les suivantes, plutôt que de continuer sur
                une sortie invalide.
        """
        if not self._steps:
            raise PipelineError("Le pipeline n'a aucune étape.")

        result = initial_input
        history: list[StepResult] = []

        for step in self._steps:
            target = await self._resolve_agent(step)
            payload = step.input_fn(result) if step.input_fn else result
            result = await self.orchestrator.submit_task(
                step.title, target, payload, timeout=step.timeout,
            )
            history.append(StepResult(title=step.title, agent=target, output=result))

        return PipelineResult(output=result, history=history)


async def fan_out(
    orchestrator,
    branches: Iterable[tuple[str, Any]],
    *,
    capabilities: Optional[dict[str, Iterable[str]]] = None,
    agents: Optional[dict[str, str]] = None,
    timeout: float = 15.0,
) -> dict[str, Any]:
    """
    Soumet plusieurs tâches en parallèle et attend toutes les réponses.

    Le pendant du pipeline séquentiel : là où `InterMeshPipeline` enchaîne,
    `fan_out` interroge plusieurs agents à la fois — utile pour agréger
    plusieurs avis avant de trancher, ou pour paralléliser un travail
    indépendant par nature.

        results = await fan_out(
            orchestrator,
            [("marché_fr", {"region": "FR"}), ("marché_de", {"region": "DE"})],
            capabilities={"marché_fr": ["market_analysis"], "marché_de": ["market_analysis"]},
        )
        results["marché_fr"]  # sortie de la branche, ou l'exception levée

    Args:
        branches:     Paires `(clé, entrée)`. La clé identifie la branche
                      dans le résultat ; elle n'est pas envoyée à l'agent.
        capabilities: Critère de découverte par clé de branche.
        agents:       Nom d'agent exact par clé de branche. Prime sur
                      `capabilities` pour cette branche.
        timeout:      Délai d'attente, commun à toutes les branches.

    Returns:
        Un dict clé → résultat. Une branche en échec y figure comme
        l'exception levée (jamais levée elle-même) : les branches saines
        restent exploitables même si une autre a échoué.
    """
    agents = agents or {}
    capabilities = capabilities or {}

    async def _run_branch(key: str, payload: Any) -> Any:
        target = agents.get(key)
        if target is None:
            found = await orchestrator.discover(capabilities=capabilities.get(key, []))
            if found.get("count", 0) == 0:
                raise PipelineError(f"Branche '{key}' : aucun agent disponible.")
            target = found["agents"][0]["name"]
        return await orchestrator.submit_task(key, target, payload, timeout=timeout)

    keys = [k for k, _ in branches]
    results = await asyncio.gather(
        *(_run_branch(k, payload) for k, payload in branches),
        return_exceptions=True,
    )
    return dict(zip(keys, results))
